File: rewrite-kotlin/src/test/java/org/openrewrite/kotlin/tree/LambdaTest.java
/*
 * Copyright 2022 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.kotlin.tree;

import org.junit.jupiter.api.Test;
import org.openrewrite.Issue;
import org.openrewrite.test.RewriteTest;

import static org.openrewrite.kotlin.Assertions.kotlin;

@SuppressWarnings("RemoveRedundantQualifierName")
class LambdaTest implements RewriteTest {

    @Test
    void binaryExpressionAsBody() {
        rewriteRun(
          kotlin(
            """
              fun method ( ) {
                  val square =  {   number    :     Int ->  number   *    number     }
              }
              """
          )
        );
    }

    @Test
    void invokedLambda() {
        rewriteRun(
          kotlin(
            """
              fun plugins ( input : ( ) -> String ) {
                println ( input( ) )
              }
              """
          )
        );
    }

    @Test
    void destructuredLambdaParams() {
        rewriteRun(
          kotlin(
            """
              abstract class SomeClass {

                  private val defaults = emptySet < String > ( )

                  abstract fun fields ( ) : List < Pair < String , Any ? > >

                  fun inputValues ( ) : List < Pair < String , Any ? > > {
                      return fields ( ) . filter { ( k , _ ) -> ! defaults . contains ( k ) }
                  }
              }
              """
          )
        );
    }

    @Test
    void multipleDestructuredLambdaParams() {
        rewriteRun(
          kotlin(
            """
              val m = mapOf(Pair("", "")).forEach { (key, value) ->
                       println(key + value)
              }
              """
          )
        );
    }

    @Issue("https://github.com/openrewrite/rewrite-kotlin/issues/60")
    @Test
    void suspendLambda() {
        rewriteRun(
          kotlin(
            """
              fun method ( ) {
                  val lambda : suspend ( ) -> Int = suspend { 1 }
              }
              """
          )
        );
    }

    @Issue("https://github.com/openrewrite/rewrite-kotlin/issues/56")
    @Test
    void suspendLambdaWithParameter() {
        rewriteRun(
          kotlin(
            """
              fun method ( ) {
                  val lambda : suspend ( Int ) -> Int = { number : Int -> number * number }
              }
              """
          )
        );
    }

    @Test
    void ignored() {
        rewriteRun(
          kotlin(
            """
              fun method() {
                  val list = listOf ( 1 , 2 , 3 )
                  list . filterIndexed { index , ignored -> index % 2 == 0 }
              }
              """
          )
        );
    }

    @Test
    void underscore() {
        rewriteRun(
          kotlin(
            """
              fun method ( ) {
                  val list = listOf ( 1 , 2 , 3 )
                  list . filterIndexed { index , _ -> index % 2 == 0 }
              }
              """
          )
        );
    }

    @Test
    void trailingComma() {
        rewriteRun(
          kotlin(
            """
              val sum: (Int, Int, ) -> Int = { x, y , -> x + y }
              """
          )
        );
    }

    @Test
    void redundantArrow() {
        rewriteRun(
          kotlin(
            """
              import java.util.function.Supplier

              class Test {
                  fun method(n: Int) {
                      val ns: Supplier<Int> = Supplier {   ->
                          n
                      }
                  }
              }
              """
          )
        );
    }

    @Issue("https://github.com/openrewrite/rewrite-kotlin/issues/110")
    @Test
    void unusedVar() {
        rewriteRun(
          kotlin(
            """
              package foo
              class Bar {
                  companion object {
                      fun bar ( e : Any ) {
                      }
                  }
              }
              """
          ),
          kotlin(
            """
              import foo.Bar
              fun test() {
                  val a = Bar . bar {
                      _ : Any? ->
                  }
              }
              """
          )
        );
    }

    @Issue("https://github.com/openrewrite/rewrite-kotlin/issues/110")
    @Test
    void fullyQualifiedKotlinTypeReference() {
        rewriteRun(
          kotlin(
            """
              package foo
              class Bar {
                  companion object {
                      fun bar ( e : Any ) {
                      }
                  }
              }
              """
          ),
          kotlin(
            """
              import foo.Bar
              fun test() {
                  val a = Bar . bar {
                      _ : kotlin . Int? ->
                  }
              }
              """
          )
        );
    }

    @Test
    void underScoreAsLamdbaParameters() {
        rewriteRun(
          kotlin(
            """
              fun method(map : HashMap<String, String>) {
                  map.forEach { (_, value) -> println("$value!") }
              }
              """
          )
        );
    }
}

File: rewrite-java/src/test/kotlin/org/openrewrite/java/tree/LambdaTest.kt
/*
 * Copyright 2020 the original author or authors.
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * https://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.openrewrite.java.tree

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.openrewrite.java.JavaParser

interface LambdaTest {

    @Test
    fun lambda(jp: JavaParser) {
        val a = jp.parse("""
            import java.util.function.Function;
            public class A {
                Function<String, String> func = (String s) -> "";
            }
        """)

        val lambda = a.classes[0].fields[0].vars[0].initializer as J.Lambda

        assertEquals(1, lambda.paramSet.params.size)
        assertTrue(lambda.body is J.Literal)
        assertEquals("(String s) -> \"\"", lambda.printTrimmed())
    }

    @Test
    fun untypedLambdaParameter(jp: JavaParser) {
        val a = jp.parse("""
            import java.util.*;
            public class A {
                List<String> list = new ArrayList<>();
                public void test() {
                    list.stream().filter(s -> s.isEmpty());
                }
            }
        """)

        assertEquals("list.stream().filter(s -> s.isEmpty())",
                a.classes[0].methods[0].body!!.statements[0].printTrimmed())
    }

    @Test
    fun optionalSingleParameterParentheses(jp: JavaParser) {
        val a = jp.parse("""
            import java.util.*;
            public class A {
                List<String> list = new ArrayList<>();
                public void test() {
                    list.stream().filter((s) -> s.isEmpty());
                }
            }
        """)

        assertEquals("list.stream().filter((s) -> s.isEmpty())",
                a.classes[0].methods[0].body!!.statements[0].printTrimmed())
    }

    @Test
    fun rightSideBlock(jp: JavaParser) {
        val a = jp.parse("""
            public class A {
                Action a = ( ) -> { };
            }

            interface Action {
                void call();
            }
        """)

        val lambda = a.classes[0].fields[0].vars[0].initializer!!
        assertEquals("( ) -> { }", lambda.printTrimmed())
    }

    @Test
    fun multipleParameters(jp: JavaParser) {
        val a = jp.parse("""
            import java.util.function.BiConsumer;
            public class A {
                BiConsumer<String, String> a = (s1, s2) -> { };
            }
        """)

        val lambda = a.classes[0].fields[0].vars[0].initializer!!
        assertEquals("(s1, s2) -> { }", lambda.printTrimmed())
    }
}
